### DISCUSSION TITLE
tutorials/intro-dgs/12-wrapping-up

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#133] [course feedback]
# tutorials/intro-dgs/12-wrapping-up

https://www.apollographql.com/tutorials/intro-dgs/12-wrapping-up

**Comment by punjprakash** [comment]
You dont want to call api /playlists/{playlist_id} if user is only wants playlist ID.  May be we can change getPayloadPlaList to 
 @DgsData(parentType="AddItemsToPlaylistPayload", field="playlist")
    public Playlist getPayloadPlaylist(DgsDataFetchingEnvironment dfe) {
        AddItemsToPlaylistPayload payload = dfe.getSource();
        Playlist playlist = payload.getPlaylist();

        // Check if the user requested any fields other than 'id' for the playlist
        List<SelectedField> requestedFields = dfe.getSelectionSet().getFields();
        if (playlist != null && requestedFields.size() > 1 && !requestedFields.contains("id")) {
            String playlistId = playlist.getId();
            return spotifyClient.playlistRequest(playlistId);
        }
        return playlist;
    }

**Reply by lizhennessy** [reply]
Hi there @punjprakash ! We've set up our datafetchers to work this way simply because of the expectations of our schema.

For example, in this `AddItemsToPlaylistPayload` type, we have a `playlist` field that we've said returns a `Playlist` object. 

```
type AddItemsToPlaylistPayload {
  "Similar to HTTP status code, represents the status of the mutation"
  code: Int!
  "Indicates whether the mutation was successful"
  success: Boolean!
  "Human-readable message for the UI"
  message: String!
  "The playlist that contains the newly added items"
  playlist: Playlist
}
```

We made the decision to use the ID for the playlist that was updated to then make a follow-up request for _all_ the playlist details because of our schema-driven approach: ideally, we're modeling our schema after the needs of our consumers, so we might imagine a scenario when the playlist is updated with additional tracks, our UI automatically updates in turn! (the previously fetched playlist would be supplemented with the new details about the tracks we just added!) This means that our schema supports queries that we expect our client applications to use, and the experiences they're trying to create. 

In this case, we expect our consumers to use the full selection of fields available on the `Playlist` object they update, to then render some change in the UI.

All that said, I love your use of the `dfe.getSelectionSet` method and I agree that's totally something that you can do!
